File: C++/JSONParser.cpp
#include "JSONParser.h"

Stock* JSONParser::parse() {
	ifstream file; string word, str;
	file.open(fileName);
	if (!file.is_open()) throw CantOpenJSONFile();
	file >> str;
	
	regex rx(".*symbol\":\"([^\"]*).*");
	smatch result;
	regex_match(str, result, rx);
	string name = result.str(1);

	regex rxx(".+?timestamp\":\\[([^\\]]*)\\].+");
	regex_match(str, result, rxx);
	string timestamps = result.str(1);
	
	regex rx1(".+?open\":\\[([^\\]]*)\\].+");
	regex_match(str, result, rx1);
	string open = result.str(1);

	regex rx2(".+?close\":\\[([^\\]]*)\\].+");
	regex_match(str, result, rx2);
	string close = result.str(1);

	regex rx3(".+?high\":\\[([^\\]]*)\\].+");
	regex_match(str, result, rx3);
	string high = result.str(1);

	regex rx4(".+?low\":\\[([^\\]]*)\\].+");
	regex_match(str, result, rx4);
	string low = result.str(1);

	Stock* s = new Stock(name);

	regex rx5("([^,]+)");
	sregex_iterator begin1(timestamps.begin(), timestamps.end(), rx5), begin2(open.begin(), open.end(), rx5), begin3(close.begin(), close.end(), rx5);
	sregex_iterator begin4(high.begin(), high.end(), rx5), begin5(low.begin(), low.end(), rx5);
	sregex_iterator end;

	while (begin1 != end) {
		smatch sm1 = *begin1; smatch sm2 = *begin2; smatch sm3 = *begin3; smatch sm4 = *begin4; smatch sm5 = *begin5;
		if (sm2.str(1) == "null") break;
		string t = sm1.str(1); double o = stof(sm2.str(1)); double c = stof(sm3.str(1)); double h = stof(sm4.str(1)); double l = stof(sm5.str(1));
		Candle* candle = new Candle(t, o, c, h, l);
		s->addCandle(candle);
		begin1++; begin2++; begin3++; begin4++; begin5++;
	}

	file.close();
	remove(fileName.c_str());
	return s;
}

double JSONParser::getCurrentPrice() {
	string str;
	fstream file(fileName);
	if (!file.is_open()) throw CantOpenJSONFile();
	file >> str;
	regex rx(".+?regularMarketPrice\":([^,]*).+");
	smatch result;
	if (!regex_match(str, result, rx))
		throw CouldntFindInfoAboutStock();
	double price = atof(result.str(1).c_str());
	file.close();
	remove(fileName.c_str());
	return price;
}

File: Java/src/stocks/JSONParser.java
package stocks;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import exceptions.*;

public class JSONParser extends Parser {
	
	public JSONParser(String fileName) {
		super(fileName);
	}

	@Override
	public Stock parse() throws CouldntFindInfoAboutStock {
		try {
		String name, timestamp, open, close, low, high;
		List<String> timestamps = new ArrayList<>(); 
		List<Double> opens = new ArrayList<>(), closes = new ArrayList<>(), lows = new ArrayList<>(), highs = new ArrayList<>();
		File file = new File(fileName);
		if (!file.exists()) throw new CantOpenJSONFile();
		BufferedReader br = new BufferedReader(new FileReader(file));
		String str = br.readLine();
		br.close(); file.delete();
		Pattern p = Pattern.compile(".*symbol\":\"([^\"]*).*");
		Matcher m = p.matcher(str);
		if(m.matches()) {
			 name = m.group(1);
		}
		else throw new CouldntFindInfoAboutStock();
		
		Pattern p1 = Pattern.compile(".*timestamp\":\\[([^\\]]*)\\].*");
		Matcher m1 = p1.matcher(str);
		if(m1.matches()) {
			 timestamp = m1.group(1);
			 Pattern p2 = Pattern.compile("([^,]+)");
			 Matcher m2 = p2.matcher(timestamp);
			 while(m2.find()) {
				 timestamps.add(m2.group(1));
			 }
		} else throw new CouldntFindInfoAboutStock();
		
		Pattern p3 = Pattern.compile(".+open\":\\[([^\\]]*)\\].+");
		Matcher m3 = p3.matcher(str);
		if(m3.matches()) {
			 open = m3.group(1);
			 Pattern p2 = Pattern.compile("([^,]+)");
			 Matcher m2 = p2.matcher(open);
			 while(m2.find()) {
				 String s = m2.group(1);
				 if(s.equals("null")) break;
				 opens.add(Double.parseDouble(m2.group(1)));
			 }
		}
		
		Pattern p4 = Pattern.compile(".+close\":\\[([^\\]]*)\\].+");
		Matcher m4 = p4.matcher(str);
		if(m4.matches()) {
			 close = m4.group(1);
			 Pattern p2 = Pattern.compile("([^,]+)");
			 Matcher m2 = p2.matcher(close);
			 while(m2.find()) {
				 String s = m2.group(1);
				 if(s.equals("null")) break;
				 closes.add(Double.parseDouble(m2.group(1)));
			 }
		}
		
		Pattern p5 = Pattern.compile(".+high\":\\[([^\\]]*)\\].+");
		Matcher m5 = p5.matcher(str);
		if(m5.matches()) {
			 high = m5.group(1);
			 Pattern p2 = Pattern.compile("([^,]+)");
			 Matcher m2 = p2.matcher(high);
			 while(m2.find()) {
				 String s = m2.group(1);
				 if(s.equals("null")) break;
				 highs.add(Double.parseDouble(m2.group(1)));
			 }
		}
		
		Pattern p6 = Pattern.compile(".+low\":\\[([^\\]]*)\\].+");
		Matcher m6 = p6.matcher(str);
		if(m6.matches()) {
			 low = m6.group(1);
			 Pattern p2 = Pattern.compile("([^,]+)");
			 Matcher m2 = p2.matcher(low);
			 while(m2.find()) {
				 String s = m2.group(1);
				 if(s.equals("null")) break;
				 lows.add(Double.parseDouble(m2.group(1)));
			 }
		}

		Stock stock = new Stock(name);
		
		for(int i = 0; i < opens.size(); i++) {
			Candle candle = new Candle(timestamps.get(i), opens.get(i), closes.get(i), highs.get(i), lows.get(i));
			stock.addCandle(candle);
		}

		return stock;
		
		} catch (IOException e) {} 
		catch (CantOpenJSONFile e) { System.out.println("Problem pri otvaranju .json fajla u kome se cuvaju podaci o akciji."); System.exit(1); }
		return null;
	}
	
	@Override
	public double getCurrentPrice() throws CouldntFindInfoAboutStock {
		try {
			double price;
			File file = new File(fileName);
			if (!file.exists()) throw new CantOpenJSONFile();
			BufferedReader br = new BufferedReader(new FileReader(file));
			String str = br.readLine();
			br.close();
			Pattern p = Pattern.compile(".*regularMarketPrice\":([^,]*).*");
			Matcher m = p.matcher(str);
			if(m.matches()) {
				price = Double.parseDouble(m.group(1));
			}
			else throw new CouldntFindInfoAboutStock();
			file.delete();
			return price;
		} catch (IOException e) {}
		catch (CantOpenJSONFile e) { System.out.println("Problem pri otvaranju .json fajla u kome se cuvaju podaci o akciji."); System.exit(1); }
		return 0;
	}
	
}
